### PR TITLE
Fix compilation on systems with Lua 5.3.0+

### LINKED ISSUE #90: lua_gettable's return type is int, not void in lua-5.3
conky-1.10.0 fails to compile against lua-5.3; the error is:

In file included from /usr/include/lua.hpp:6:0,
                 from /build/conky/src/conky-1.10.0/src/luamm.hh:30,
                 from /build/conky/src/conky-1.10.0/src/luamm.cc:24:
/build/conky/src/conky-1.10.0/src/luamm.cc: In member function 'void lua::state::gettable(int)':
/build/conky/src/conky-1.10.0/src/luamm.cc:370:3: error: no matches converting function 'safe_misc_trampoline' to type 'lua_CFunction {aka int (*)(struct lua_State*)}'
   lua_pushcfunction(cobj.get(), (&safe_misc_trampoline<&lua_gettable, 1>));
   ^
/build/conky/src/conky-1.10.0/src/luamm.cc:168:7: note: candidate is: template<void (* misc)(lua_State*, int), int nresults> int lua::{anonymous}::safe_misc_trampoline(lua_State*)
   int safe_misc_trampoline(lua_State *l)
       ^
src/CMakeFiles/conky.dir/build.make:675: recipe for target 'src/CMakeFiles/conky.dir/luamm.cc.o' failed

It turns out lua_gettable returns an int in lua-5.3. The attached patch fixes this issue, though as currently written it is specific to lua-5.3.

### PR DESCRIPTION
**Pull request by marcpayne**
Since Lua 5.3.0, the return type of `lua_gettable` and similar functions is int, rather than void. Thus, `safe_misc_trampoline<&lua_gettable, 1>` does not compile (the first template parameter is a pointer to function with return type void). A solution was provided by vesath in issue #90, which I modified slightly with Lua version checks. It compiles and works on my system, but I can't test with older Lua versions. If travis can compile it, we should be good.

Fixes #90 and half of #100.

### CODE REVIEW

**Comment by Nooby4Ever** [comment]
The build flags corresponding to the: cairo, imlib2 and rsvg will cause errors as stated in #116 . So you have to disable them when Lua 5.3 is used, this should also be clearly documented somewhere to avoid confusion.

**Comment by marcpayne** [comment]
TL;DR: I will probably close this pull request. Some input on my last code snippet below is welcome. In the meantime, I'll look at the C++ standards.

---

Yeah, until someone can update the bindings you mention (which I can't at this point), `ConkyPlatformChecks.cmake` could be changed to output an error if the user tries to build them with any version of Lua past 5.1. This should probably also be mentioned in the LUA API section of the man page.

Regarding `pkg_search_module(...)`, the only reason it works is that it contains `lua>=5.1`. This will work on any distro that ships a pkgconfig file named `lua.pc`, which most distros should. However, if a user installs a Lua 5.2 package alongside 5.3 on a distro that ships, say, `lua-5.2.pc`, then conky will only see Lua 5.2. I don't know if this is intentional. The following might work better:

```
pkg_search_module(LUA REQUIRED lua>=5.1 lua53 lua5.3 lua-5.3 lua52 lua5.2 lua-5.2 lua51 lua5.1 lua-5.1)
```

Finally, after sleeping on it, I realized that my patch would be much simpler if I just overloaded the existing `safe_misc_trampoline` function template. No messy `#if` statements needed.

```
template<void (*misc)(lua_State *, int), int nresults>
int safe_misc_trampoline(lua_State *l)
{
	misc(l, 1);
	return nresults;
}

template<int (*misc)(lua_State *, int), int nresults>
int safe_misc_trampoline(lua_State *l)
{
	misc(l, 1);
	return nresults;
}
```

Like I often say, this "works for me." Input is welcome before I close this pull request, particularly because I don't know the finer details of overloading function templates with different function pointer parameters, and whether this code will work for everyone else.

**Comment by Nooby4Ever** [comment]
The idea of overloading should work in theory, but i have never used it before so i unfortunately can not comment on the code itself.

**Comment by antonio-malcolm** [comment]
I think tolua++ still needs the older 5.1 Lua, so end users need the legacy library for the dependency. Are there any plans to move to another means of binding to Lua, to output to Cairo? Neither tolua and tolua++ have been updated in YEARS (though tolua, miraculously, compiles against the latest verssion of Lua)..
